### ISSUE TITLE
How to escalate issues?

### ISSUE THREAD

**Issue by beorn** [electron/governance#536]
There's a lot of frustration seen in https://github.com/electron/electron/issues/21457 - It seems like an issue that's been allowed to fester for way too long without an satisfactory answer from the Electron team.  It could do with some escalation... but to where?  Is there a working group that deals with roadmap and perhaps coordinating sponsored feature work, if that's required?

**Comment by MarshallOfSound** [comment]
> It seems like an issue that's been allowed to fester for way too long without a satisfactory answer from the Electron team.

Electron is an open source project that's available for everyone to use for free. We don't provide commercial support for the project, and don't have a customer SLA either. Beyond the engineers you've interacted with, there's no higher power to inform—everyone who has opinions on the issue or the power to make changes are already aware of the GitHub issue.

Please understand also that we're also still looking at the problem space regularly, both through technical discussions internally and by replying in the GitHub issue.

> Is there a working group that deals with roadmap

No, but we try to share news on public-facing initiatives whenever we can!

> perhaps coordinating sponsored feature work

For ESM, the fundamental issue is not money, or time, or sponsorship. The issues have been outlined multiple times in that thread but I'll say them once more here again for good measure.

* There is currently no clear path to enabling ESM as an entry point to an Electron app in a way that is 100% compatible with how you do certain things in Electron apps nowadays
* There **are** ways to do it by making significant architectural changes/additions, but there is no need to do those given they would effectively be unnecessarily codifying the "workarounds" that are mentioned in the issue you linked (that we plan to document soon)

If a technical proposal arrives that may solve the problem altogether, we can have a discussions on the tradeoffs of the approach for existing users and discuss a path forward. Note that said discussion still belongs in the original issue.

There is also a good case to be made for documenting "how to use ESM" better because you can definitely use it today, but that discussion also lives in the issue you linked, not here.

**Comment by beorn** [comment]
Yeah I agree summarizing the ESM issues, plans, and workarounds (an Electron ESM FAQ?) somewhere else so it can be linked to (and updated as discussions evolve) is probably better than rehashing them again and again on a thread with hundreds of comments.

I do think a roadmap could be motivational and many projects have those, but I understand everything is of course on a completely volunteer basis — people contribute what they want.

**Comment by ilg-ul** [comment]
> > Is there a working group that deals with roadmap
> No, but we try to share news on public-facing initiatives whenever we can!

I don't know how the project is governed, but wouldn't it be useful to have such a working group (Electron ESM WG)? With the first goal to define such a roadmap, for example.

**Comment by erickzhao** [comment]
Hey @ilg-ul, thanks for stopping by the repo. You might have already seen the ESM entry point PR (https://github.com/electron/electron/pull/37535), but I didn't want to leave you hanging.

To answer your question directly: I don't think it would be useful to add an additional working group here for a few reasons.

1. Our governance model uses WGs as a structure for long-lasting work for core maintainers, which doesn't really fit this use case. There isn't really a long-lasting initiative here that would be broken down into milestones with a roadmap and all that jazz, and the administrative overhead of adding another group just wouldn't be worth it.
2. Also, while we have 8 working groups already, the same ~20 full-time contributors are usually part of 3 or more WGs at a time (and often have non-Electron job responsibilities). Adding more groups doesn't mean more resources will be allocated to the project because we are already a high-leverage core team that's stretched fairly thin.

**Comment by ilg-ul** [comment]
Hi @erickzhao, thank you for your answer.

If the mentioned PR will be enough for the VS Code team to allow VS Code extensions to import ES modules, indeed there will be need for a new WG.

Otherwise... I don't know what are your priorities, but I hope that this one will get a bit more attention than it got in the last few years.

**Comment by erickzhao** [comment]
I think this is veering off-topic for this issue but I can answer a bit to your follow-ups as well:

> If the mentioned PR will be enough for the VS Code team to allow VS Code extensions to import ES modules

I can't pretend to understand how it works, but I hope this does unblock some technical limitation with how extensions are loaded as well. Ultimately, it will be up to the VSCode team.

> Otherwise... I don't know what are your priorities, but I hope that this one will get a bit more attention than it got in the last few years.

If you want to understand _why_ this wasn't a P0 "drop all things and work on this" issue, here are a few reasons:

* ESM via dynamic import in CJS is supported in Node.js (and Electron). As per (https://github.com/microsoft/vscode/issues/130367#issuecomment-1030721392), this seems to be broken for VSCode extensions for VSCode-specific reasons.
```diff
- const prettyMs = require('pretty-ms')
+ const prettyMs = (await import('pretty-ms')).default
```
* Dual publishing npm modules as CJS and ESM is a thing (we even do it for certain packages like [`@electron/osx-sign`](https://github.com/electron/osx-sign/blob/main/package.json)). If individual ecosystem maintainers choose to only support ESM, it's their conscious decision rather than Node.js forcing them to choose one option or the other.

Regardless, maintainers have been going back to the drawing board with this problem every once in a while and occasionally engaging in discussion with the community in the GH issue. It just can't monopolize our time because there's a lot more that goes into keeping Electron a secure and delightful experience for developers using the framework.

I don't think it's fair to say that we've been ignoring this issue. The problem was always a difficult architectural challenge, and it turns out that a maintainer finally figured out a clever solution. IMO, this is really a "do we have the solution, or do we not" thing and there isn't much news to communicate until there is.

PRs have also always been welcome from external contributors as well, but I don't think we've ever gotten a proposal from anyone that would have made any progress here.

**Comment by ilg-ul** [comment]
Thank you @erickzhao for the follow-ups.

> If individual ecosystem maintainers choose to only support ESM, it's their conscious decision rather than Node.js forcing them to choose one option or the other.

Yes, it is their conscious decision, but the ESM route is a one way street. Once one of your dependencies made the move to ESM only, you can continue to use an old release for a while, but at a certain moment you'll need an update and this will make your package ESM only too. So yes, a decision, but with only one choice. :-(

In practice, the faster this transition happens, the less pain. 

In my case, due to maintenance issues, I had to update my Node.js packages, only to revert to CJS those that are consumed by VS Code too. :-(

Let's hope that with this new PR things will finally move forward.
